Sources/SwiftyLLVM/Context.swift

```diff
@@ -0,0 +1,73 @@
+internal import llvmc
+
+/// An opaque object that owns core LLVM data structures.
+///
+/// A single context is not thread safe but different contexts can execute on different threads
+/// simultaneously.
+public struct Context {
+
+  /// A handle to the LLVM object wrapped by this instance.
+  let llvm: LLVMContextRef
+
+  /// Returns the result of calling `action` wiht a new LLVM context.
```

Comment:
`Context.withNew { ... }` reads quite oddly to me, like it's doing something with a new closure.  A more obvious spelling for me would be to construct a context and then have the context do something, e.g. `Context().do {  ... }`.  Is there a reason such an interface is inappropriate?

Reply:
With `Context().do { ... }` it becomes less clear that one is only supposed to use the context in the closure. The API should remind us that the lifetime of the context is only valid in some explicit scope.

For example:
```
// This is okay:
Context.withNew { (llvm) in
  let t = llvm.i64
  print(t)
}

// This isn't:
let llvm = Context()
let t = llvm.i64
print(t)
```

With `Context().do { ... }`I'm not even sure Swift wouldn't drop the receiver of the call. AFAIK there is no way to call a mutating function on a rvalue so I suspect the optimizer would be free to assume that the result of `Context()` is unused once we evaluated the function denoted by `do`. At the very least, it would be allowed to make a copy and then call `Context`'s deinitializer at any point, sending us back to UB.

The alternative is the PR you proposed, where each handle to a LLVM resource whose lifetime depends on the context keeps a reference to that context. Or we can keep the `withXXX` approach and come up with better names, obviously.

Reply:
I don't think `withNew` is much help.  You also are only supposed to use `t` inside that closure and nothing reminds you of that.

> At the very least, it would be allowed to make a copy and then call Context's deinitializer at any point, sending us back to UB.

`Context` very obviously (to me) needs to be a Swift `class`.  That's the appropriate construct for the semantics you actually get.  There's not much worse than unsafe reference semantics masquerading as value semantics.  You at least get safe reference semantics if you use classes.

Incidentally, speaking of names `context` would be a much better name than `llvm` for that parameter.

Reply:
> I don't think withNew is much help. You also are only supposed to use t inside that closure and nothing reminds you of that.

I disagree. There's a note in the API docs and there is precedent in Swift for `withXXX` having the requirement that the argument cannot escape.

> There's not much worse than unsafe reference semantics masquerading as value semantics. You at least get safe reference semantics if you use classes.

But here I don't think we do, actually. Let's continue this discussion in a live meeting.

I am also not sure `context` is a much better name. This reference is more or less a handle to an instance of the LLVM library. In contrast `context` is mostly a meaningless word, especially because it is heavily overloaded in compilers.

Reply:
> I am also not sure context is a much better name. This reference is more or less a handle to an instance of the LLVM library. In contrast context is mostly a meaningless word, especially because it is heavily overloaded in compilers.

But this isn't a compiler; this is an LLVM wrapper library.  Given that knowledge, think about what the two words `llvm` and `context` tell the reader.